src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,583 @@
+                        "  {1} {0} {2}ms", uri, response.StatusCode.ToString(), sw.ElapsedMilliseconds.ToString()));
+                    return new HttpSourceResult();
+                }
+
+                response.EnsureSuccessStatusCode();
+
+                await CreateCacheFile(result, response, cacheContext, cancellationToken);
+
+                log.LogInformation(string.Format(CultureInfo.InvariantCulture,
+                    "  {1} {0} {2}ms", uri, response.StatusCode.ToString(), sw.ElapsedMilliseconds.ToString()));
+
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Wraps logging of the initial request and throttling.
+        /// This method does not use the cache.
+        /// </summary>
+        private async Task<HttpResponseMessage> GetAsync(
+            HttpRequestMessage request,
+            ILogger log,
+            CancellationToken cancellationToken)
+        {
+            log.LogInformation(string.Format(CultureInfo.InvariantCulture, "  {0} {1}.",
+                request.Method.Method.ToUpperInvariant(),
+                request.RequestUri));
+
+            // Read the response headers before reading the entire stream to avoid timeouts from large packages.
+            Func<Task<HttpResponseMessage>> throttledRequest = () => SendWithCredentialSupportAsync(
+                    request,
+                    HttpCompletionOption.ResponseHeadersRead,
+                    cancellationToken);
+
+            var response = await GetThrottled(throttledRequest, log);
+
+            return response;
+        }
+
+        private static async Task<HttpResponseMessage> GetThrottled(Func<Task<HttpResponseMessage>> request, ILogger log)
+        {
+            if (_throttle == null)
+            {
+                return await request();
+            }
+            else
+            {
+                try
+                {
+                    await _throttle.WaitAsync();
+
+                    log.LogVerbose($"Current http requests queued: {_throttle.CurrentCount + 1}");
+
+                    return await request();
+                }
+                finally
+                {
+                    _throttle.Release();
+                }
+            }
+        }
+
+        public Task<HttpResponseMessage> GetAsync(Uri uri, ILogger log, CancellationToken token)
+        {
+            var request = new HttpRequestMessage(HttpMethod.Get, uri);
+
+            return GetAsync(request, log, token);
+        }
+
+        public async Task<Stream> GetStreamAsync(Uri uri, ILogger log, CancellationToken token)
+        {
+            var response = await GetAsync(uri, log, token);
+            response.EnsureSuccessStatusCode();
+
+            return await response.Content.ReadAsStreamAsync();
+        }
+
+        public async Task<JObject> GetJObjectAsync(Uri uri, ILogger log, CancellationToken token)
+        {
+            using (var stream = await GetStreamAsync(uri, log, token))
```

Comment:
this feels a little like a duplicate of Utils.GetResource, do we need both of them?